### DISCUSSION TITLE
Best way to ensure procedure types are JSON-serializable?

### DISCUSSION THREAD

**Discussion by huw** [trpc/trpc#2571] [Q&A]
I hunted around the issues and couldn’t quite find anything, so I just wanted to make sure I’m not missing something here.

I have a procedure that returns a type (`Temporal` from `temporal-polyfill`) that can be serialized to JSON, but can’t be deserialized without a custom transformer. Right now, I’m not so much interested in writing a transformer—but I at least want to make sure that when I call my procedure on the client, TypeScript infers the type correctly. Right now, TypeScript thinks it’s a `Temporal`, but it’s definitely a `string` (as Temporal supports `toJSON()` and serializes to a string).

Right now I just make sure to type my procedures in the router like so, using `Jsonify` from `type-fest`:

```ts
const router = trpc.router({
  hello: trpc.procedure.query(
    async (): Promise<Jsonify<Awaited<ReturnType<typeof backendFunction>>>> =>
      backendFunction()
  ),
});
```

Am I missing an obvious way to get this behaviour automatically?

**Comment by KATT** [comment]
A few strategies:

A) Use superjson and add a temporal serializer (I do exactly this in my app)
B) Make a helper util for `Jsonify` 


### A) superjson helper for temporal

```ts
import { Temporal } from '@js-temporal/polyfill';
import superjson from 'superjson';

// Adds ability to deserialize/serialize `Temporal.PlainDate` to JSON
superjson.registerCustom<Temporal.PlainDate, string>(
  {
    isApplicable: (v): v is Temporal.PlainDate =>
      v instanceof Temporal.PlainDate,
    serialize: (v) => v.toJSON(),
    deserialize: (v) => Temporal.PlainDate.from(v),
  },
  'Temporal.PlainDate',
);

// Adds ability to deserialize/serialize `Temporal.Instant` to JSON
superjson.registerCustom<Temporal.Instant, string>(
  {
    isApplicable: (v): v is Temporal.Instant => v instanceof Temporal.Instant,
    serialize: (v) => v.toJSON(),
    deserialize: (v) => Temporal.Instant.from(v),
  },
  'Temporal.Instant',
);

export const transformer = superjson;
```

### B) `jsonify` util

```ts
// something like this
const jsonify = <T>(obj: T /* TODO narrow to Jsonify-compat somehow */): Jsonify<T> => obj as any;
```
